browser/ui/ai_chat/print_preview_extractor_unittest.cc

```diff
@@ -306,6 +368,74 @@ class PrintPreviewExtractorTest : public ChromeRenderViewHostTestHarness {
     VerifyPrintPreviewUIState();
   }
 
+  void SimulateFullFlow(bool extractor_error) {
+    auto full_pdf_region = CreatePageRegion(64).region;
+    internal_extractor()->SetPreviewPageTextExtractorForTesting(
+        std::make_unique<MockPreviewPageTextExtractor>(
+            full_pdf_region.Duplicate(), extractor_error));
+
+    // Simulate page composition
+    internal_extractor()->OnCompositePdfPageDone(
+        0, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+        CreatePageRegion(8).region);
+    internal_extractor()->OnCompositePdfPageDone(
+        1, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+        CreatePageRegion(16).region);
+    internal_extractor()->OnCompositeToPdfDone(
+        0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+        full_pdf_region.Duplicate());
+  }
+
+  template <typename T>
+  void RunTestCase(base::test::TestFuture<T>& future,
+                   const std::string& mime_type,
+                   bool use_capture_pdf,
+                   const std::string& expected_error_msg,
+                   bool extractor_error,
+                   bool simulate_partial_composition) {
+    const bool expect_error = !expected_error_msg.empty();
+    if (!expect_error || extractor_error) {
+      SimulateFullFlow(extractor_error);
+    } else if (simulate_partial_composition) {
+      // Simulate partial composition without OnCompositeToPdfDone
+      internal_extractor()->OnCompositePdfPageDone(
+          0, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+          CreatePageRegion(8).region);
+      internal_extractor()->OnCompositePdfPageDone(
+          1, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+          CreatePageRegion(16).region);
+      internal_extractor()->OnPreviewReady();
+    } else {
+      // For missing preview data case, just trigger preview ready
+      internal_extractor()->OnPreviewReady();
```

Comment:
OnPreviewReady should be a private method, and we should call OnCompositeToPdfDone with error status to match what would happen in production code.
Don't simulate partial composition without OnCompositeToPdfDone if it's impossible.

Reply:
Actually, after looking into print_render_frame_helper from Chromium more, it does seems possible that only OnCompositePdfPageDone is called and OnCompositeToPdfDone is not called.
It is possible that DidPreviewPage was called, then PrintPreviewFailed, without MetafileReadyForPrinting being called.
So if we want to, maybe we could simulate this without manually calling OnPreviewReady but calling the PrintPreviewFailed instead?